bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,940 @@
+     *
+     * It implements a CompletableFuture by means of which the caller can wait for the response data to come in. And
+     * which, in the case of fatal errors, gets completed exceptionally.
+     *
+     * It handles the following fatal error events by notifying the owner..
+     *
+     * <li>onHeaders() HTTP unauthorised codes</li>
+     *
+     * @author Andrew Fiddian-Green - Initial Contribution
+     */
+    private static abstract class BaseAdapter extends Stream.Listener.Adapter implements AdapterErrorHandler {
+
+        protected final Clip2Bridge owner;
+        protected final List<String> strings = new ArrayList<>();
+        protected final CompletableFuture<String> completable = new CompletableFuture<>();
+
+        protected BaseAdapter(Clip2Bridge owner) {
+            this.owner = owner;
+        }
+
+        @Override
+        public void fatalError(Error error) {
+            Exception e;
+            if (Error.UNAUTHORISED.equals(error)) {
+                e = new IllegalAccessException("HTTP 2.0 request not authorized");
+            } else {
+                e = new ApiException("HTTP 2.0 stream " + error.toString().toLowerCase());
+            }
+            completable.completeExceptionally(e);
+            owner.fatalError(this, error);
+        }
+
+        /**
+         * Check the reply headers to see whether the request was authorised.
+         */
+        @Override
+        public void onHeaders(@Nullable Stream stream, @Nullable HeadersFrame frame) {
+            Objects.requireNonNull(frame);
+            MetaData metaData = frame.getMetaData();
+            if (metaData.isResponse()) {
+                int httpStatus = ((Response) metaData).getStatus();
+                switch (httpStatus) {
+                    case HttpStatus.UNAUTHORIZED_401:
+                    case HttpStatus.FORBIDDEN_403:
+                        fatalError(Error.UNAUTHORISED);
+                    default:
+                }
+            }
+        }
+    }
+
+    /**
+     * Adapter for regular HTTP GET/PUT request stream events.
+     *
+     * It assembles the incoming text data into an HTTP 'content' entity. And when the last data frame arrives, it
+     * returns the full content by completing the CompletableFuture with that data.
+     *
+     * In addition to those handled by the parent, it handles the following fatal error events by notifying the owner..
+     *
+     * <li>onIdleTimeout()</li>
+     * <li>onTimeout()</li>
+     *
+     * @author Andrew Fiddian-Green - Initial Contribution
+     */
+    private static class ContentAdapter extends BaseAdapter {
+
+        protected ContentAdapter(Clip2Bridge owner) {
+            super(owner);
+        }
+
+        @Override
+        public void onData(@Nullable Stream stream, @Nullable DataFrame frame, @Nullable Callback callback) {
+            Objects.requireNonNull(frame);
+            Objects.requireNonNull(callback);
```

Comment:
These can never happen. I am just using them to make the Maven compiler potential NPE warnings go away. (the code `Objects.requireNonNull()` is the same as `if (x != null)`..

Reply:
To be precise, when used like this, it's the same as something like `if (x == null) throw new NullPointerException()` which has some merit in order to fail fast.

As I see this is a general approach used, I may need to get back to this. Other contributors / @openhab/add-ons-maintainers  might also have something to add.

Reply:
I also use `Objects.requireNonNull` simply as a hint to the compiler that I expect this thing to never be null, even though it can't statically deduce that itself.